{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "913feaae-5220-43a7-a5a1-9e2d860829f2",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Imports and data loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3828732d-1c2c-496e-b8ee-1c506aadf32f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "import glob\n",
    "from tqdm import tqdm\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf1ae869-d6ec-4033-ae32-c577813d1821",
   "metadata": {},
   "source": [
    "## Set output directory of `gather`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "de226719-910b-4bde-9380-080a702b2b5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "outdir = '/Users/olgabot/botryllus/adhoc-analysis/2022-apr--gather-botryllus-in-human-mouse-with-kmers/'\n",
    "# ! mkdir $outdir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3fae891c-8615-4557-b983-cf7bafcad7b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 23996176\n",
      "drwxr-xr-x  32001 olgabot  staff   1.0M May 13 10:55 .\n",
      "drwxr-xr-x      4 olgabot  staff   128B Apr 16 13:47 ..\n",
      "-rw-r--r--      1 olgabot  staff    85K Apr 16 14:48 BHF.csv\n",
      "-rw-r--r--      1 olgabot  staff   701M May 13 10:44 botryllus_gather_mouse_human_results.parquet\n",
      "-rw-r--r--      1 olgabot  staff    31K Apr 16 14:49 g1-dot-t1_frame-colon-1.csv\n",
      "-rw-r--r--      1 olgabot  staff    11K Apr 16 14:50 g10-dot-t1_frame-colon-1.csv\n",
      "-rw-r--r--      1 olgabot  staff    98K Apr 16 17:59 g100-dot-t1_frame-colon-1.csv\n",
      "-rw-r--r--      1 olgabot  staff    20K Apr 16 23:44 g1000-dot-t1_frame-colon-1.csv\n",
      "-rw-r--r--      1 olgabot  staff    46K Apr 20 15:09 g10000-dot-t1_frame-colon-3.csv\n"
     ]
    }
   ],
   "source": [
    "! ls -lha $outdir | head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ea5f0847-c9bb-4bed-9ee7-59fd2917fca8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>i_query</th>\n",
       "      <th>kmer_query</th>\n",
       "      <th>kmer_hp</th>\n",
       "      <th>hashval</th>\n",
       "      <th>name_query</th>\n",
       "      <th>i_found</th>\n",
       "      <th>kmer_found</th>\n",
       "      <th>name_found</th>\n",
       "      <th>n_kmers</th>\n",
       "      <th>intersect_bp</th>\n",
       "      <th>containment</th>\n",
       "      <th>symbol</th>\n",
       "      <th>species</th>\n",
       "      <th>found_i</th>\n",
       "      <th>homolog_group</th>\n",
       "      <th>genomic_coord</th>\n",
       "      <th>all_homologs_found</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>183</td>\n",
       "      <td>IKSLTYNHSFEIESQFDEREIKKQ</td>\n",
       "      <td>hpphphppphphppphpppphppp</td>\n",
       "      <td>2231394216540723782</td>\n",
       "      <td>g34.t1 frame:1</td>\n",
       "      <td>425</td>\n",
       "      <td>LTDPSVSQQINPRNTIDTKDYSTH</td>\n",
       "      <td>ENST00000506763.5 gene_id=ENSG00000214367.9;tr...</td>\n",
       "      <td>10</td>\n",
       "      <td>50</td>\n",
       "      <td>0.05618</td>\n",
       "      <td>HAUS3</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>HAUS3---Haus3</td>\n",
       "      <td>Chr4:2231836-2242164(-)</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>184</td>\n",
       "      <td>KSLTYNHSFEIESQFDEREIKKQS</td>\n",
       "      <td>pphphppphphppphpppphpppp</td>\n",
       "      <td>10588571181114802221</td>\n",
       "      <td>g34.t1 frame:1</td>\n",
       "      <td>426</td>\n",
       "      <td>TDPSVSQQINPRNTIDTKDYSTHR</td>\n",
       "      <td>ENST00000506763.5 gene_id=ENSG00000214367.9;tr...</td>\n",
       "      <td>10</td>\n",
       "      <td>50</td>\n",
       "      <td>0.05618</td>\n",
       "      <td>HAUS3</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>HAUS3---Haus3</td>\n",
       "      <td>Chr4:2231836-2242164(-)</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>185</td>\n",
       "      <td>SLTYNHSFEIESQFDEREIKKQSG</td>\n",
       "      <td>phphppphphppphpppphpppph</td>\n",
       "      <td>12040406379061937897</td>\n",
       "      <td>g34.t1 frame:1</td>\n",
       "      <td>427</td>\n",
       "      <td>DPSVSQQINPRNTIDTKDYSTHRL</td>\n",
       "      <td>ENST00000506763.5 gene_id=ENSG00000214367.9;tr...</td>\n",
       "      <td>10</td>\n",
       "      <td>50</td>\n",
       "      <td>0.05618</td>\n",
       "      <td>HAUS3</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>HAUS3---Haus3</td>\n",
       "      <td>Chr4:2231836-2242164(-)</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>186</td>\n",
       "      <td>LTYNHSFEIESQFDEREIKKQSGA</td>\n",
       "      <td>hphppphphppphpppphpppphh</td>\n",
       "      <td>14509062212459467365</td>\n",
       "      <td>g34.t1 frame:1</td>\n",
       "      <td>428</td>\n",
       "      <td>PSVSQQINPRNTIDTKDYSTHRLY</td>\n",
       "      <td>ENST00000506763.5 gene_id=ENSG00000214367.9;tr...</td>\n",
       "      <td>10</td>\n",
       "      <td>50</td>\n",
       "      <td>0.05618</td>\n",
       "      <td>HAUS3</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>HAUS3---Haus3</td>\n",
       "      <td>Chr4:2231836-2242164(-)</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>187</td>\n",
       "      <td>TYNHSFEIESQFDEREIKKQSGAR</td>\n",
       "      <td>phppphphppphpppphpppphhp</td>\n",
       "      <td>7876319629203808642</td>\n",
       "      <td>g34.t1 frame:1</td>\n",
       "      <td>429</td>\n",
       "      <td>SVSQQINPRNTIDTKDYSTHRLYQ</td>\n",
       "      <td>ENST00000506763.5 gene_id=ENSG00000214367.9;tr...</td>\n",
       "      <td>10</td>\n",
       "      <td>50</td>\n",
       "      <td>0.05618</td>\n",
       "      <td>HAUS3</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>HAUS3---Haus3</td>\n",
       "      <td>Chr4:2231836-2242164(-)</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   i_query                kmer_query                   kmer_hp  \\\n",
       "0      183  IKSLTYNHSFEIESQFDEREIKKQ  hpphphppphphppphpppphppp   \n",
       "1      184  KSLTYNHSFEIESQFDEREIKKQS  pphphppphphppphpppphpppp   \n",
       "2      185  SLTYNHSFEIESQFDEREIKKQSG  phphppphphppphpppphpppph   \n",
       "3      186  LTYNHSFEIESQFDEREIKKQSGA  hphppphphppphpppphpppphh   \n",
       "4      187  TYNHSFEIESQFDEREIKKQSGAR  phppphphppphpppphpppphhp   \n",
       "\n",
       "                hashval      name_query  i_found                kmer_found  \\\n",
       "0   2231394216540723782  g34.t1 frame:1      425  LTDPSVSQQINPRNTIDTKDYSTH   \n",
       "1  10588571181114802221  g34.t1 frame:1      426  TDPSVSQQINPRNTIDTKDYSTHR   \n",
       "2  12040406379061937897  g34.t1 frame:1      427  DPSVSQQINPRNTIDTKDYSTHRL   \n",
       "3  14509062212459467365  g34.t1 frame:1      428  PSVSQQINPRNTIDTKDYSTHRLY   \n",
       "4   7876319629203808642  g34.t1 frame:1      429  SVSQQINPRNTIDTKDYSTHRLYQ   \n",
       "\n",
       "                                          name_found  n_kmers  intersect_bp  \\\n",
       "0  ENST00000506763.5 gene_id=ENSG00000214367.9;tr...       10            50   \n",
       "1  ENST00000506763.5 gene_id=ENSG00000214367.9;tr...       10            50   \n",
       "2  ENST00000506763.5 gene_id=ENSG00000214367.9;tr...       10            50   \n",
       "3  ENST00000506763.5 gene_id=ENSG00000214367.9;tr...       10            50   \n",
       "4  ENST00000506763.5 gene_id=ENSG00000214367.9;tr...       10            50   \n",
       "\n",
       "   containment symbol species  found_i  homolog_group  \\\n",
       "0      0.05618  HAUS3   human        0  HAUS3---Haus3   \n",
       "1      0.05618  HAUS3   human        0  HAUS3---Haus3   \n",
       "2      0.05618  HAUS3   human        0  HAUS3---Haus3   \n",
       "3      0.05618  HAUS3   human        0  HAUS3---Haus3   \n",
       "4      0.05618  HAUS3   human        0  HAUS3---Haus3   \n",
       "\n",
       "             genomic_coord  all_homologs_found  \n",
       "0  Chr4:2231836-2242164(-)               False  \n",
       "1  Chr4:2231836-2242164(-)               False  \n",
       "2  Chr4:2231836-2242164(-)               False  \n",
       "3  Chr4:2231836-2242164(-)               False  \n",
       "4  Chr4:2231836-2242164(-)               False  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(os.path.join(outdir, 'g34-dot-t1_frame-colon-1.csv'))\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "13ea6365-0b4c-4d08-932f-4c787a446d09",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  1%|█                                                                                                                 | 305/31972 [00:01<02:47, 189.46it/s]/Users/olgabot/opt/miniconda3/envs/botryllus/lib/python3.10/site-packages/IPython/core/magics/execution.py:1335: DtypeWarning: Columns (16) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code, glob, local_ns)\n",
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 31972/31972 [04:20<00:00, 122.55it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19342484, 18)\n",
      "CPU times: user 4min 3s, sys: 36.4 s, total: 4min 39s\n",
      "Wall time: 5min 4s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>i_query</th>\n",
       "      <th>kmer_query</th>\n",
       "      <th>kmer_hp</th>\n",
       "      <th>hashval</th>\n",
       "      <th>name_query</th>\n",
       "      <th>i_found</th>\n",
       "      <th>kmer_found</th>\n",
       "      <th>name_found</th>\n",
       "      <th>n_kmers</th>\n",
       "      <th>intersect_bp</th>\n",
       "      <th>containment</th>\n",
       "      <th>symbol</th>\n",
       "      <th>species</th>\n",
       "      <th>found_i</th>\n",
       "      <th>homolog_group</th>\n",
       "      <th>genomic_coord</th>\n",
       "      <th>all_homologs_found</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>150</td>\n",
       "      <td>YSFTLKDDTGEVVLDQWNKASLVP</td>\n",
       "      <td>hphphpppphphhhpphpphphhh</td>\n",
       "      <td>3018107519398277786</td>\n",
       "      <td>g25414.t1 frame:1</td>\n",
       "      <td>1238</td>\n",
       "      <td>PKISLRRSSLKYLGCRYSEIKPYG</td>\n",
       "      <td>ENST00000361833.7 gene_id=ENSG00000152092.16;t...</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>0.011583</td>\n",
       "      <td>ASTN1</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>ASTN1---Astn1</td>\n",
       "      <td>Chr1:176855118-177164904(-)</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000361833.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>151</td>\n",
       "      <td>SFTLKDDTGEVVLDQWNKASLVPG</td>\n",
       "      <td>phphpppphphhhpphpphphhhh</td>\n",
       "      <td>1234479950528962468</td>\n",
       "      <td>g25414.t1 frame:1</td>\n",
       "      <td>1239</td>\n",
       "      <td>KISLRRSSLKYLGCRYSEIKPYGL</td>\n",
       "      <td>ENST00000361833.7 gene_id=ENSG00000152092.16;t...</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>0.011583</td>\n",
       "      <td>ASTN1</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>ASTN1---Astn1</td>\n",
       "      <td>Chr1:176855118-177164904(-)</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000361833.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>152</td>\n",
       "      <td>FTLKDDTGEVVLDQWNKASLVPGK</td>\n",
       "      <td>hphpppphphhhpphpphphhhhp</td>\n",
       "      <td>2563348012698350512</td>\n",
       "      <td>g25414.t1 frame:1</td>\n",
       "      <td>1240</td>\n",
       "      <td>ISLRRSSLKYLGCRYSEIKPYGLD</td>\n",
       "      <td>ENST00000361833.7 gene_id=ENSG00000152092.16;t...</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>0.011583</td>\n",
       "      <td>ASTN1</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>ASTN1---Astn1</td>\n",
       "      <td>Chr1:176855118-177164904(-)</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000361833.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15</td>\n",
       "      <td>NFHVGVKFVSSPKRGGVSQYLKGI</td>\n",
       "      <td>phphhhphhpphpphhhpphhphh</td>\n",
       "      <td>1046426320533182525</td>\n",
       "      <td>g25414.t1 frame:1</td>\n",
       "      <td>576</td>\n",
       "      <td>SVQPIVKLVSTATTAPPSTAPSGP</td>\n",
       "      <td>ENST00000453269.7 gene_id=ENSG00000106290.15;t...</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>0.015038</td>\n",
       "      <td>TAF6</td>\n",
       "      <td>human</td>\n",
       "      <td>1</td>\n",
       "      <td>TAF6---Taf6</td>\n",
       "      <td>Chr7:100107070-100119858(-)</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000453269.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16</td>\n",
       "      <td>FHVGVKFVSSPKRGGVSQYLKGIF</td>\n",
       "      <td>hphhhphhpphpphhhpphhphhh</td>\n",
       "      <td>2154375574397474662</td>\n",
       "      <td>g25414.t1 frame:1</td>\n",
       "      <td>577</td>\n",
       "      <td>VQPIVKLVSTATTAPPSTAPSGPG</td>\n",
       "      <td>ENST00000453269.7 gene_id=ENSG00000106290.15;t...</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>0.015038</td>\n",
       "      <td>TAF6</td>\n",
       "      <td>human</td>\n",
       "      <td>1</td>\n",
       "      <td>TAF6---Taf6</td>\n",
       "      <td>Chr7:100107070-100119858(-)</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000453269.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   i_query                kmer_query                   kmer_hp  \\\n",
       "0      150  YSFTLKDDTGEVVLDQWNKASLVP  hphphpppphphhhpphpphphhh   \n",
       "1      151  SFTLKDDTGEVVLDQWNKASLVPG  phphpppphphhhpphpphphhhh   \n",
       "2      152  FTLKDDTGEVVLDQWNKASLVPGK  hphpppphphhhpphpphphhhhp   \n",
       "3       15  NFHVGVKFVSSPKRGGVSQYLKGI  phphhhphhpphpphhhpphhphh   \n",
       "4       16  FHVGVKFVSSPKRGGVSQYLKGIF  hphhhphhpphpphhhpphhphhh   \n",
       "\n",
       "               hashval         name_query  i_found                kmer_found  \\\n",
       "0  3018107519398277786  g25414.t1 frame:1     1238  PKISLRRSSLKYLGCRYSEIKPYG   \n",
       "1  1234479950528962468  g25414.t1 frame:1     1239  KISLRRSSLKYLGCRYSEIKPYGL   \n",
       "2  2563348012698350512  g25414.t1 frame:1     1240  ISLRRSSLKYLGCRYSEIKPYGLD   \n",
       "3  1046426320533182525  g25414.t1 frame:1      576  SVQPIVKLVSTATTAPPSTAPSGP   \n",
       "4  2154375574397474662  g25414.t1 frame:1      577  VQPIVKLVSTATTAPPSTAPSGPG   \n",
       "\n",
       "                                          name_found  n_kmers  intersect_bp  \\\n",
       "0  ENST00000361833.7 gene_id=ENSG00000152092.16;t...        3            15   \n",
       "1  ENST00000361833.7 gene_id=ENSG00000152092.16;t...        3            15   \n",
       "2  ENST00000361833.7 gene_id=ENSG00000152092.16;t...        3            15   \n",
       "3  ENST00000453269.7 gene_id=ENSG00000106290.15;t...        2            10   \n",
       "4  ENST00000453269.7 gene_id=ENSG00000106290.15;t...        2            10   \n",
       "\n",
       "   containment symbol species  found_i  homolog_group  \\\n",
       "0     0.011583  ASTN1   human        0  ASTN1---Astn1   \n",
       "1     0.011583  ASTN1   human        0  ASTN1---Astn1   \n",
       "2     0.011583  ASTN1   human        0  ASTN1---Astn1   \n",
       "3     0.015038   TAF6   human        1    TAF6---Taf6   \n",
       "4     0.015038   TAF6   human        1    TAF6---Taf6   \n",
       "\n",
       "                 genomic_coord all_homologs_found      transcript_id  \n",
       "0  Chr1:176855118-177164904(-)              False  ENST00000361833.7  \n",
       "1  Chr1:176855118-177164904(-)              False  ENST00000361833.7  \n",
       "2  Chr1:176855118-177164904(-)              False  ENST00000361833.7  \n",
       "3  Chr7:100107070-100119858(-)              False  ENST00000453269.7  \n",
       "4  Chr7:100107070-100119858(-)              False  ENST00000453269.7  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "csvs = glob.glob(os.path.join(outdir, \"*.csv\"))\n",
    "dfs = []\n",
    "\n",
    "for csv in tqdm(csvs):\n",
    "    df = pd.read_csv(csv)\n",
    "    # Convert hashval to string\n",
    "    df['hashval'] = df['hashval'].astype(str)\n",
    "    df['transcript_id'] = df['name_found'].str.split().str[0]\n",
    "    dfs.append(df)\n",
    "gather_results = pd.concat(dfs)\n",
    "print(gather_results.shape)\n",
    "gather_results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7b3db1ac-45df-4f32-aa1f-12f32a833731",
   "metadata": {},
   "outputs": [],
   "source": [
    "parquet = os.path.join(outdir, \"botryllus_gather_mouse_human_results.parquet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "cea71587-c492-478a-9f50-e9b81afbee93",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 38 s, sys: 46.4 s, total: 1min 24s\n",
      "Wall time: 2min\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "gather_results.to_parquet(\n",
    "    parquet\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6db07124-9533-484a-b496-d10e830f1854",
   "metadata": {},
   "outputs": [],
   "source": [
    "del gather_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "429e11dc-9ad7-4838-8613-758e4efe00bf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:botryllus]",
   "language": "python",
   "name": "conda-env-botryllus-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
